### PR TITLE
APIGOV-17423 - Integrate generic agent resources

### PR DESCRIPTION
**Pull request by kflan-io**

### CODE REVIEW

**Comment by kflan-io** [comment]
@dfeldick @jcollins-axway @sbolosan @vivekschauhan

File: pkg/agent/agent.go [hunk 1]
```diff
@@ -72,6 +72,14 @@ type agentData struct {
 
 var agent = agentData{}
 
+// getAgentTypes - return the agent types
```
Comment (jcollins-axway) [inline pkg/agent/agent.go:75]: This is just returning a value, can't it be a variable?  Also, can you use the AgentType enum rather then the int representation?

File: pkg/agent/agent.go [hunk 2]
```diff
@@ -315,6 +323,10 @@ func cleanUp() {
 
 // GetAgentResourceType - Returns the Agent Resource path element
 func getAgentResourceType() string {
+	// Set resource for Generic Type
+	if AgentResourceType == "" {
+		AgentResourceType = getAgentTypes()[int(agent.cfg.AgentType)]
```
Comment (jcollins-axway) [inline pkg/agent/agent.go:328]: With changes suggested above, this casting is not necessary

File: pkg/agent/edgeagents.go [hunk 3]
```diff
@@ -16,13 +17,27 @@ func edgeDiscoveryAgent(res *apiV1.ResourceInstance) *v1alpha1.EdgeDiscoveryAgen
...
 func createEdgeDiscoveryAgentStatusResource(status, message string) *v1alpha1.EdgeDiscoveryAgent {
-	agentRes := v1alpha1.EdgeDiscoveryAgent{}
+	agentRes := edgeDiscoveryAgent(GetAgentResource())
 	agentRes.Name = agent.cfg.GetAgentName()
 	agentRes.Status.Version = config.AgentVersion
 	agentRes.Status.State = status
 	agentRes.Status.Message = message
 
-	return &agentRes
+	dataplaneType := "Edge"
```
Comment (jcollins-axway) [inline pkg/agent/edgeagents.go:26]: Make this a constant?

File: pkg/agent/edgeagents.go [hunk 4]
```diff
@@ -47,6 +62,20 @@ func createEdgeTraceabilityAgentStatusResource(status, message string) *v1alpha1
 	agentRes.Status.State = status
 	agentRes.Status.Message = message
 
+	dataplaneType := "Edge"
```
Comment (jcollins-axway) [inline pkg/agent/edgeagents.go:65]: reuse constant from above, package level constant?

File: pkg/agent/agent.go [hunk 5]
```diff
@@ -369,16 +381,36 @@ func updateAgentStatus(status, message string) error {
...
-		_, err = agent.apicClient.ExecuteAPI(coreapi.PUT, subResURL, nil, buffer)
-		if err != nil {
-			return err
+		if genResource != nil {
+			err := updateAgentStatusAPI(resource, agentResourceType)
+			if err != nil {
+				log.Warn("Could not update the generic agent reference")
+			}
 		}
+		return updateAgentStatusAPI(resource, agentResourceType)
```
Comment (jcollins-axway) [inline pkg/agent/agent.go:399]: if genResource != nil we update the status twice? Is 394 meant to be genResource?

**Review by jcollins-axway — CHANGES_REQUESTED** [review]

File: pkg/agent/awsagents.go [hunk 6]
```diff
@@ -22,6 +23,20 @@ func createAWSDiscoveryAgentStatusResource(status, message string) *v1alpha1.AWS
 	agentRes.Status.State = status
 	agentRes.Status.Message = message
 
+	dataplaneType := "AWS"
```
Comment (jcollins-axway) [inline pkg/agent/awsagents.go:26]: Same as Edge

File: pkg/agent/awsagents.go [hunk 7]
```diff
@@ -47,6 +62,20 @@ func createAWSTraceabilityAgentStatusResource(status, message string) *v1alpha1.
...
```
Comment (jcollins-axway) [inline pkg/agent/awsagents.go:65]: Same as Edge

**Review by jcollins-axway — CHANGES_REQUESTED** [review]

File: pkg/agent/agent.go (hunk 1 above)
Comment (kflan-io) [reply pkg/agent/agent.go:75]: with no const for a map, I thought returning the map would be safer in the SDK; I am happy to change it.

File: pkg/agent/agent.go (hunk 2 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:328]: Its a private map, no?

File: pkg/agent/agent.go (hunk 1 above)
Comment (jcollins-axway) [reply pkg/agent/agent.go:75]: Its a private map, no?

Comment (kflan-io) [reply pkg/agent/agent.go:75]: done

**Review by vivekschauhan — APPROVED** [review]

**Review by jcollins-axway — APPROVED** [review]

**Review by sbolosan — APPROVED** [review]

**Pull request merged**